blog/_posts/2021-05-18-argocd-private-clusters.md

```diff
@@ -0,0 +1,414 @@
+
+Luckily for us, we can configure everything in a declarative way.
+
+First, create a service account manually in the target cluster for Argo CD.
+
+``` yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: argocd-manager
+  namespace: kube-system
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: argocd-manager-role
+rules:
+- apiGroups:
+  - '*'
+  resources:
+  - '*'
+  verbs:
+  - '*'
+- nonResourceURLs:
+  - '*'
+  verbs:
+  - '*'
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: argocd-manager-role-binding
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: argocd-manager-role
+subjects:
+- kind: ServiceAccount
+  name: argocd-manager
+  namespace: kube-system
+```
+
+For Argo CD, cluster configuration and credentials are stored in secrets which must have a label `argocd.argoproj.io/secret-type: cluster`.
+
+The secret data must include the following fields:
+
+* name: a name for the cluster
+* server: the cluster API URL, here we should use the internal URL of the data plane
+* namespaces: optional comma-separated list of namespaces that are accessible in that cluster
+* config: JSON representation of the cluster credentials
+
+More information on how this structure looks like can be found in the [Argo CD documentation](https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#clusters).
+
+The following script is an example of how to will fetch the bearer token for the service account created earlier and create such a secret for our target cluster.
+
+``` bash
+export TARGET_CTX=gke-eu1
+export ARGOCD_CTX=argocd
+
+name=$(kubectl get --context $TARGET_CTX sa -n kube-system argocd-manager -o jsonpath='{.secrets[0].name}')
+
+ca=$(kubectl get --context $TARGET_CTX -n kube-system secret/$name -o jsonpath='{.data.ca\.crt}')
+token=$(kubectl get --context $TARGET_CTX -n kube-system secret/$name -o jsonpath='{.data.token}' | base64 --decode)
+namespace=$(kubectl get --context $TARGET_CTX -n kube-system secret/$name -o jsonpath='{.data.namespace}' | base64 --decode)
+
+cat <<EOF | kubectl apply --context $ARGOCD_CTX -n argocd -f -
+apiVersion: v1
+kind: Secret
+metadata:
+  name: gke-eu1
+  labels:
+    argocd.argoproj.io/secret-type: cluster
+type: Opaque
+stringData:
+  name: gke-eu1
+  server: https://gke-eu1.inlets:443
+  config: |
+    {
+      "bearerToken": "${token}",
+      "tlsClientConfig": {
+        "insecure": true
+      }
+    }
+EOF
+```
+
+Notice the `"insecure" : true"`? This is probably necessary as the certificate of the Kubernetes API service is not valid for the domain name `gke-eu1.inlets`. Where we have access to kubeadm or k3s for the private cluster, we can add a TLS SAN name, and the solution works by directly tunnelling the API server.
```

Comment:
You could be more reassuring here.

Also, link to / mention inlets-connect - the proxy that can mean "insecure" isn't needed